text/0249-v2-experiments.md

```diff
@@ -0,0 +1,380 @@
+  depend on a package containing unstable APIs, or with `0.x` for the version.
+- If a third-party package depends on an unstable API in a non-obvious way (for
+  example, only in the implementation of a construct, not in its public API),
+  that might break for customers when upgrading to a version of `aws-cdk-lib`
+  that has broken that functionality compared to the `aws-cdk-lib` version the
+  third-party construct is built against (basically, the same scenario from
+  above that explains why we can no longer have unstable code in stable mono-CDK
+  modules). None of the options solve the problem of allowing third-party
+  libraries to safely depend on experimental Construct Library code; however,
+  the fact that all experimental code in this variant is shipped in
+  `aws-cdk-lib` makes this particular problem more likely to manifest itself.
+- Graduating a module to stable will be a breaking change for customers. We can
+  mitigate this downside by keeping the old unstable module around, but that
+  leads to duplicated classes in the same package.
+
+### Option 2: separate single unstable package
+
+Instead of vending the unstable modules together with the stable ones, we can
+vend a second mono-CDK, `aws-cdk-lib-experiments` (actual name can be changed
+before release of course). A customer will have to explicitly depend on
+`aws-cdk-lib-experiments`, which will be released in version `0.x` to make it
+even more obvious that this is unstable code. `aws-cdk-lib-experiments` would
+have a caret peer dependency on `aws-cdk-lib`.
+
+So, the above Cognito example would look like this:
+
+```ts
+import * as cognito from 'aws-cdk-lib/aws-cognito';
+import * as cognito_preview from 'aws-cdk-lib-experiments/aws-cognito';
+
+const idp = new cognito_preview.UserPoolIdentityProviderOidc(this, 'OIDC', { ... });
+const supported = [cognito.UserPoolClientIdentityProvider.custom("MyProviderName")];
+const userPoolClient = new cognito.UserPoolClient(...);
+```
+
+Advantages:
+
+- Very explicit (customer has to add a dependency on a package with
+  "experiments" in the name and version `0.x`).
+- It's possible for unstable modules to depend on other unstable modules.
+
+Disadvantages:
+
+- It's not possible for stable modules to depend on unstable ones. This has
+  serious side implications:
+  - All unstable modules that have stable dependents today will have to be
+    graduated before `v2.0` is released.
+  - Before a module is graduated, all of its dependencies need to be graduated.
+  - It will not be possible to add new dependencies on unstable modules to
+    stable modules in the future (for example, that's a common need for
+    StepFunction Tasks).
+- Graduating a module to stable will be a breaking change for customers. We can
+  mitigate this downside by keeping the old unstable module around, but that
+  leads to duplicated classes.
+
+### Option 3: separate multiple unstable packages
```

Comment:
I like this option. The constraints (disadvantages) feel reasonable.

> We can mitigate this downside by keeping the old unstable package around, but that leads to duplicated classes.

Re: duplicate classes - we can completely delete the code off the code base and mark the released packages as deprecated, so no code duplication. Existing customers can continue to use them but won't get newer features.